### PR TITLE
Node tree rewrite

### LINKED ISSUE #44: Node children lifetime
It appears to be counter-obvious that we have `parent.add(&child)` yet it's not parent holding the child alive but the other way around. In some cases, this goes in the way of encapsulation: e.g. when we load a number of different things from a file, we may want to return a single group that everything is attached to, yet today if you do this - all the children will get removed for losing the references.

There does not appear to be an easy way to address this, on the technical level. The implementation largely prefers storing the parents only and working with them. Ideas:
  1. Keep the children locally by groups and scenes. Problem: hard to synchronize with those children changing parents.
  2. Have `children: Vec<Pointer<Node>>>` alongside `parent: WeakPointer<Node>` in the nodes. Problem: traverse performance, given that accessing children array is expensive, and accessing the parent requires conversion to a strong pointer...
  3. Leave the implementation as is, but change the API to `child.attach(&parent)` to make the relationship clear.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #44

Core of the change is switching relationships from "child->parent" to "parent->first_child->sibling", which gives us the following benefits:
  - clean children lifetimes, as expected by users
    - no need to store all the meshes returned by gltf/obj loaders: perhaps, we can follow-up with a PR that allows querying meshes by name from their parents, so that we don't even return all the meshes directly in the loader API
  - nice `Group::add(&something)`, like Three-js does
  - make `Scene` truly unique and node-less
  - no need for unique scene IDs
    ~~- technically, objects can now be present in multiple scenes at once~~
  - immutable node data upon traversal (!):
    - we just compute the world visibility/transform on the fly
    - thus, no need for a separate `update_transforms` step
    - uses pretty iterators ;)
    - as a downside, not as straightforward to reflect the world info to the user (see TODO)

Has other changes making things nicer:
  - refactored frame render code
  - no `&mut self` on objects
  - idiomatic hub indexing
  - safer group/camera/text/audio spawning (internal)
  - more robust message handling (panic on unprocessed/unexpected messages)
  - can debug print the `Operation`

TODO:
- [x] check/fix docs
- [x] compute world transform/visibility for the user if needed (is it needed?)
- [x] provide API to detach from parent
- [x] reviews!
- [x] profile the cost of recomputing the world transforms

### CODE REVIEW

File: src/hub.rs [hunk 1]
```diff
@@ -185,24 +148,55 @@ impl Hub {
...
+                    }
+                    _ => unreachable!()
                 },
+            };
+        }
+
+        for (child_ptr, sibling) in deferred_sibling_updates {
+            let child = &mut self.nodes[&child_ptr];
+            if child.next_sibling.is_some() {
+                error!("Attaching a child that still has an old parent, discarding siblings");
```
Comment (vitvakatu) [inline src/hub.rs]: How we can change parent now?

Comment (kvark) [reply src/hub.rs]: Oh right, I forgot about this :) Now added to TODO list for this PR.
Basically, we can either merge the old siblings with the new ones (like `Scene::add` does), or require the user to call `Group::remove(&item)` before adding to another group. Either way, the `remove` needs to be added.

File: src/hub.rs [hunk 2]
```diff
@@ -185,24 +148,55 @@ impl Hub {
...
-                    *mat = material;
                 },
-                Operation::SetTexelRange(base, size) => if let SubNode::Visual(ref mut material, _) = node.sub_node {
-                    match *material {
-                        material::Material::Sprite(ref mut params) => params.map.set_texel_range(base, size),
-                        _ => panic!("Unsupported material for texel range request"),
+                Operation::AddChild(child_ptr) => match node.sub_node {
+                    SubNode::Group { ref mut first_child } => {
+                        let sibling = mem::replace(first_child, Some(child_ptr.clone()));
+                        deferred_sibling_updates.push((child_ptr, sibling));
```
Comment (alteous) [inline src/hub.rs]: Interesting.

File: src/hub.rs [hunk 3]
```diff
@@ -240,40 +234,83 @@ impl Hub {
...
         mesh: &DynamicMesh,
     ) {
-        match self.get_mut(&mesh).sub_node {
+        match self[mesh].sub_node {
             SubNode::Visual(_, ref mut gpu_data) => gpu_data.pending = Some(mesh.dynamic.clone()),
             _ => unreachable!(),
         }
     }
+
+    pub(crate) fn walk(&self, base: &Option<NodePointer>) -> TreeWalker {
```
Comment (alteous) [inline src/hub.rs]: 👍

File: src/node.rs [hunk 4]
```diff
@@ -72,28 +74,10 @@ impl From<TransformInternal> for Transform {
...
-    pub transform: Transform,
-    /// World transform (relative to the world's origin).
-    pub world_transform: Transform,
     /// Is `Node` visible by cameras or not?
     pub visible: bool,
-    /// The same as `visible`, used internally.
-    pub world_visible: bool,
+    /// Relative to parent transform.
+    pub transform: Transform,
     /// Material in case this `Node` has it.
```
Comment (alteous) [inline src/node.rs]: Was this ever useful?

File: src/render/mod.rs [hunk 5]
```diff
@@ -654,13 +653,9 @@ impl Renderer {
...
             let p: [[f32; 4]; 4] = camera.matrix(self.aspect_ratio()).into();
-            let node = &hub.nodes[&camera.object.node];
-            let w = match node.scene_id {
-                Some(id) if Some(id) == scene_id => node.world_transform,
-                Some(_) => panic!("Camera does not belong to this scene"),
-                None => node.transform,
-            };
-            let mx_view = Matrix4::from(w.inverse_transform().unwrap());
+            let node = &hub[&camera];
+            let world_transform = node.transform; //TODO!!!
```
Comment (alteous) [inline src/render/mod.rs]: What's this for?

File: src/scene.rs [hunk 6]
```diff
@@ -4,15 +4,13 @@ use object;
...
-use object::Object;
+use node::{Node, NodePointer};
+use object::{Base, Object};
 use texture::{CubeMap, Texture};
 
+use std::mem;
 use std::sync::MutexGuard;
 
-/// Unique identifier for a scene.
-pub type Uid = usize;
```
Comment (alteous) [inline src/scene.rs]: 👍

File: src/hub.rs (hunk 2 above)
Comment (kvark) [reply src/hub.rs]: borrow checker...

File: src/node.rs (hunk 4 above)
Comment (kvark) [reply src/node.rs]: wait, why is `Material` here? O_O

Comment (alteous) [reply src/node.rs]: Why not?!

Comment (kvark) [reply src/node.rs]: Ah, this is user-exposed node, not the internal one.
Yeah, let it be. The whole reflection thing (with `sync`) I'd consider non-idiomatic, so we shouldn't be too eager to optimize it.

File: src/render/mod.rs (hunk 5 above)
Comment (kvark) [reply src/render/mod.rs]: TODO: compute the world transform for the camera, which is expected here

**Comment by kvark** [comment]
Alright, I rebased the code and implemented the missing bits.
Only the profiling is left, plus maybe a few more looks, esp in the new code.
Note: unfortunately, there is a bit of duplication atm between Group reacting to events and Scene implementation. I'm sure we can make it nicer.

**Comment by kvark** [comment]
Here are the profiling results (note: I only used `callgrind_annotate`, so don't have that much insight at the moment):
  1. `descend` does show up near the top of the profile
  2. cause likely being large amount of copying (and a bit of `stack` reallocation). This should be partially solved by future improvements in Rust itself, some possible workarounds from our side, or caching the transforms at least.
  3. `group` examples with 20K instances runs OK on my U-series CPU, so I think we are fine for now.

Looks like I addressed all the action items from my side. Waiting for more reviews now (@alteous , @vitvakatu ).

**Comment by kvark** [comment]
I believe in the long term our architecture should work like this:
- `Renderer` only walks the tree once, for each node
  - figures out what render bins does an object fall into (e.g.: main screen, shadow for light X, off-screen buffer, etc)
  - builds the GPU instance blocks for each bin (separately) and appends them to lists to be directly uploaded to GPU constant buffer memory
  - writes down GPU commands in the appropriate command buffers

TL;DR: there is no strong reason for `Renderer` to walk the tree multiple times. We can make rendering much faster on the CPU side, and tree traversal is not (going to be) a bottleneck.

File: examples/group.rs [hunk 7]
```diff
@@ -102,28 +102,41 @@ fn create_cubes(
     list
 }
 
-const COLORS: [three::Color; 6] = [0xffff80, 0x8080ff, 0x80ff80, 0xff8080, 0x80ffff, 0xff80ff];
-
-const SPEEDS: [f32; 6] = [0.7, -1.0, 1.3, -1.6, 1.9, -2.2];
+struct LevelDesc {
+    color: three::Color,
+    speed: f32,
```
Comment (alteous) [inline examples/group.rs]: What is the unit of this? I assume radians/second?

File: Cargo.toml [hunk 8]
```diff
@@ -28,7 +31,7 @@ cgmath = { version = "0.15", features = ["mint"] }
 derivative = "1.0"
 froggy = "0.4.4"
 genmesh = "0.5"
-gfx = "0.16"
+gfx = "0.16.3"
```
Comment (vitvakatu) [inline Cargo.toml]: Why?

File: src/hub.rs [hunk 9]
```diff
@@ -240,40 +272,94 @@ impl Hub {
...
+        let mut walker = TreeWalker {
+            hub: self,
+            only_visible,
+            stack: Vec::with_capacity(default_stack_size),
+        };
+        walker.descend(base);
+        walker
+    }
+
+    pub(crate) fn walk(&self, base: &Option<NodePointer>) -> TreeWalker {
```
Comment (alteous) [inline src/hub.rs]: Nit: Perhaps `walk_visible`?

File: examples/shapes.rs [hunk 10]
```diff
@@ -7,34 +7,40 @@ use three::Object;
...
         let material = three::material::Wireframe { color: 0xFF0000 };
         win.factory.mesh(geometry, material)
     };
     msphere.set_position([-3.0, 3.0, 0.0]);
-    msphere.set_parent(&win.scene);
+    win.scene.add(&msphere);
 
-    let mut mline = {
+    // test removal from scene
+    win.scene.remove(&mcyl);
```
Comment (vitvakatu) [inline examples/shapes.rs]: Is it supposed to be here?

File: src/hub.rs (hunk 9 above)
Comment (vitvakatu) [reply src/hub.rs]: As in froggy, it's better to use shorter names.

File: Cargo.toml (hunk 8 above)
Comment (kvark) [reply Cargo.toml]: Because your instancing changes require it

File: examples/group.rs (hunk 7 above)
Comment (kvark) [reply examples/group.rs]: correct

File: examples/shapes.rs (hunk 10 above)
Comment (kvark) [reply examples/shapes.rs]: yes - it's a small test for object removal. May not be the best place, but better than nothing

File: src/hub.rs (hunk 9 above)
Comment (kvark) [reply src/hub.rs]: it's for internal use only anyhow. Most of the time we are only concerned about visible nodes

File: Cargo.toml (hunk 8 above)
Comment (vitvakatu) [reply Cargo.toml]: I'm not a fan of strict version rules in Cargo. It will use the latest available version anyway.

File: src/hub.rs (hunk 9 above)
Comment (vitvakatu) [reply src/hub.rs]: Yes, I remember your strategy :smile:
alt

File: Cargo.toml (hunk 8 above)
Comment (kvark) [reply Cargo.toml]: I think your position is in the minority within the community then. Latest will be used on fresh checkout, if we don't check in `Cargo.lock`. But people who already have `Cargo.lock` will get breakage if you don't write the exact version in the `Cargo.toml`. Another case is - some crate that depends on you may say `gfx = "=0.16.2"` and it will try to build but will fail even after `cargo update`.

TL;DR: Cargo versioning rules are here to stay and be respected

Comment (kvark) [reply Cargo.toml]: Also, there is a proposal (which I supported) about making `cargo publish` checking out all the **minimal** compatible versions (as opposed to latest). Think about it ;)

Comment (vitvakatu) [reply Cargo.toml]: Okay, you convicted me

**Comment by kvark** [comment]
bors r-
Almost forgot to fix the text!!!

**Comment by kvark** [comment]
bors r=vitvakatu
